File: src/factories/payment.factory.ts
import { PaymentService } from '../dependencies/payment'
import { CreditService } from '../services/credit.service'
import { StripeService } from '../services/stripe.service'

export type PaymentType = 'credit' | 'card'

export const paymentFactory = (paymentType: PaymentType): PaymentService => {
  if (paymentType === 'credit') return new CreditService()
  if (paymentType === 'card') return new StripeService()
  throw new Error('Invalid payment type')
}

File: rust_test/src/factory/payment_factory.rs
use crate::{
    dependencies::payment::PaymentService,
    services::{credit_payment::CreditPaymentService, stripe_payment::StripePaymentService},
};

pub enum PaymentType {
    Credits,
    Card,
}

pub fn payment_factory(payment_type: PaymentType) -> Box<dyn PaymentService> {
    match payment_type {
        PaymentType::Credits => Box::new(CreditPaymentService::new()),
        PaymentType::Card => Box::new(StripePaymentService::new()),
    }
}
